### PR TITLE
Implement MediaList interface

### PR DESCRIPTION
**Pull request by canova**
Added MediaList interface and implemented `mediaText`, `length` and `index` attributes.
r? @Manishearth 

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors

- [x] There are tests for these changes

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14433)

### CODE REVIEW

File: components/script/dom/medialist.rs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+                           MediaListBinding::Wrap)
+    }
+
+}
+
+impl MediaListMethods for MediaList {
+    // https://drafts.csswg.org/cssom/#dom-medialist-mediatext
+    fn MediaText(&self) -> DOMString {
+        let mut s = String::new();
+        self.media_queries.read().to_css(&mut s).unwrap();
```
Comment (Manishearth) [inline components/script/dom/medialist.rs]: I think there is a `to_css_string()` method that does this for you

**Review by Manishearth — COMMENTED** [review]
Looking good, but to be able to test this we need `CSSMediaRule`'s `media` attribute to work

File: components/script/dom/medialist.rs (hunk 1 above)
Comment (canaltinova) [inline components/script/dom/medialist.rs]: Oh, I was thinking that this kind of a method would be good 😄

**Comment by Manishearth** [comment]
```

  ▶ Unexpected subtest result in /_mozilla/mozilla/interfaces.html:
  │ FAIL [expected PASS] Interfaces exposed on the window
  │   → assert_true: If this is failing: DANGER, are you sure you want to expose the new interface MediaList to all webpages as a property on the global? Do not make a change to this file without review from jdm or Ms2ger for that specific change! expected true got false
```

You will need to add `MediaList` to the `mozilla/interfaces.html` test (somewhere in `tests/wpt/mozilla`)

File: components/script/dom/medialist.rs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        // Step 2
+        if let Err(_) = m {
+            return;
+        }
+        // Step 3
+        let m_serialized = m.unwrap().to_css_string();
+        let mut media_list = self.media_queries.write();
+        let queries = media_list.media_queries.clone();
+        let iter = queries.iter().enumerate();
+        for (i, query) in iter {
```
Comment (Manishearth) [inline components/script/dom/medialist.rs]: This is `O(n^2)`

```rust
let new_vec = media_list.media_queries.drain(..).filter(|q| q.to_css_string() == m_serialized).collect();
*media_list.media_queries = new_vec;
```

File: components/script/dom/medialist.rs [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        let mut parser = Parser::new(&medium);
+        let m = MediaQuery::parse(&mut parser);
+        // Step 2
+        if let Err(_) = m {
+            return;
+        }
+        // Step 3
+        let m_serialized = m.clone().unwrap().to_css_string();
+        self.media_queries.read().media_queries.iter().map(|query| {
+            if m_serialized == query.to_css_string() {
```
Comment (Manishearth) [inline components/script/dom/medialist.rs]: Map is lazy, and will not execute in this case.

You want to use `self.media_queries.read().media_queries.iter().any(|q| m_serialized == q.to_css_string())` and return early if it's false.

File: components/script/dom/medialist.rs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    }
+
+    // https://drafts.csswg.org/cssom/#dom-medialist-item
+    fn Item(&self, index: u32) -> Option<DOMString> {
+        self.media_queries.read().media_queries.get(index as usize)
+                                            .and_then(|query| {
+            let mut s = String::new();
+            query.to_css(&mut s).unwrap();
+            Some(DOMString::from_string(s))
+
```
Comment (Manishearth) [inline components/script/dom/medialist.rs]: nit: extra line

**Comment by canaltinova** [comment]
Added a wpt test.

**Comment by Manishearth** [comment]
@bors-servo r+

Nice test!

**Comment by Manishearth** [comment]
Bunch of CSStest pass expected fails that we missed

http://build.servo.org/builders/linux-rel-css/builds/1262/steps/test/logs/stdio

**Comment by canaltinova** [comment]
Oh, nice :) Updated the test expectations. Should I remove my custom test?

**Comment by Manishearth** [comment]
Nah, leave it in. The more the merrier.

**Comment by Manishearth** [comment]
The CSS tests don't cover everything here anyway. They're missing a lot of cssom stuff.

**Comment by canaltinova** [comment]
Okay, it's good now I guess.

**Pull request merged**